test/socket-io.test.ts

```diff
@@ -0,0 +1,189 @@
+  }
+
+  beforeAll(async () => {
+    server = await startServer(createServer);
+  });
+
+  afterEach(() => {
+    client?.close();
+    client = null;
+  });
+
+  afterAll(async () => {
+    client?.close();
+    await server.close();
+  });
+
+  it('should connect', (done) => {
+    const socket = socketio('http://localhost:5000', {
+      path: '/graphql',
+      transports: ['websocket'],
+    });
+
+    socket.on('connect', () => {
+      socket.close();
+      done();
+    });
+  }, 10000);
+
+  it('should subscribe', async () => {
+    const socket = await createClient(
+      graphql`
+        subscription TestTodoUpdatedSubscription(
+          $input: TodoUpdatedSubscriptionInput!
+        ) {
+          todoUpdated(input: $input) {
+            todo {
+              text
+            }
+          }
+        }
+      `,
+      {
+        input: {
+          id: '1',
+        },
+      },
+    );
+
+    await socket.authenticate();
+
+    expect(
+      await socket.getSubscriptionResult({
+        topic: `todo:1:updated`,
+        data: {
+          id: '1',
+          text: 'Make work',
+        },
+      }),
+    ).toMatchInlineSnapshot(`
+      Object {
+        "event": "subscription update",
+        "payload": Object {
+          "data": Object {
+            "todoUpdated": Object {
+              "todo": Object {
+                "text": "Buy a unicorn",
+              },
+            },
+          },
+          "id": "foo",
+        },
+      }
+    `);
+  });
+
+  it('should unsubscribe', async () => {
+    const socket = await createClient(
+      graphql`
+        subscription TestTodoUpdatedSubscription(
+          $input: TodoUpdatedSubscriptionInput!
+        ) {
+          todoUpdated(input: $input) {
+            todo {
+              text
+            }
+          }
+        }
+      `,
+      {
+        input: {
+          id: '1',
+        },
+      },
+    );
+
+    await socket.authenticate();
+    await socket.subscribe();
+
+    expect(server.subscriber._queues.size).toEqual(1);
+
+    await socket.unsubscribe();
+
+    expect(server.subscriber._queues.size).toEqual(0);
+
+    await Promise.all([socket.subscribe(), socket.unsubscribe()]);
+
+    expect(server.subscriber._queues.size).toEqual(0);
+  });
+
+  it.only('should not race', async (done) => {
+    const socket = await createClient(
+      graphql`
+        subscription TestTodoUpdatedSubscription(
+          $input: TodoUpdatedSubscriptionInput!
+        ) {
+          todoUpdated(input: $input) {
+            todo {
+              text
+            }
+          }
+        }
+      `,
+      {
+        input: {
+          id: '1',
+        },
+      },
+    );
+
+    await socket.authenticate();
+
+    const range = Array.from({ length: 2 }, (_, i) => i);
+    const promises = [] as any[];
+    for (const id of range) {
+      promises.push(socket.subscribe(`s-${id}`));
+      promises.push(socket.unsubscribe(`s-${id}`));
+
+      await sleep();
```

Comment:
not that I can reproduce this with a single pair of calls, but it's not 100% consistent. doing 2 iterations seems to ensure it shows up